src/drivers/intel/cavs/ssp.c

```diff
@@ -962,6 +963,174 @@ static int ssp_get_fifo(struct dai *dai, int direction, int stream_id)
 	return dai->plat_data.fifo[direction].offset;
 }
 
+/* Functions for HW timestamp */
+
+static inline uint32_t ssp_ts_local_tsctrl_addr(int index)
+{
+#if CONFIG_APOLLOLAKE
+	/* TSCTRL registers for SSP0, 1, 2, and 3 are in continuous
+	 * registers space while SSP4 and more are handled with other
+	 * macro.
+	 */
+	if (index < DAI_NUM_SSP_BASE)
+		return TIMESTAMP_BASE + TS_I2S_LOCAL_TSCTRL(index);
+	else
+		return TIMESTAMP_BASE + TS_I2SE_LOCAL_TSCTRL(index);
+#else
+	return TIMESTAMP_BASE + TS_I2S_LOCAL_TSCTRL(index)
+#endif
+}
+
+static inline uint32_t ssp_ts_local_offs_addr(int index)
+{
+#if CONFIG_APOLLOLAKE
+	if (index < DAI_NUM_SSP_BASE)
+		return TIMESTAMP_BASE + TS_I2S_LOCAL_OFFS(index);
+	else
+		return TIMESTAMP_BASE + TS_I2SE_LOCAL_OFFS(index);
+#else
+	return TIMESTAMP_BASE + TS_I2S_LOCAL_OFFS(index)
+#endif
+}
+
+static inline uint32_t ssp_ts_local_sample_addr(int index)
+{
+#if CONFIG_APOLLOLAKE
+	if (index < DAI_NUM_SSP_BASE)
+		return TIMESTAMP_BASE + TS_I2S_LOCAL_SAMPLE(index);
+	else
+		return TIMESTAMP_BASE + TS_I2SE_LOCAL_SAMPLE(index);
+#else
+	return TIMESTAMP_BASE + TS_I2S_LOCAL_SAMPLE(index)
+#endif
+}
+
+static inline uint32_t ssp_ts_local_walclk_addr(int index)
+{
+#if CONFIG_APOLLOLAKE
+	if (index < DAI_NUM_SSP_BASE)
+		return TIMESTAMP_BASE + TS_I2S_LOCAL_WALCLK(index);
+	else
+		return TIMESTAMP_BASE + TS_I2SE_LOCAL_WALCLK(index);
+#else
+	return TIMESTAMP_BASE + TS_I2S_LOCAL_WALCLK(index)
+#endif
+}
+
+static inline uint32_t ssp_ts_tscc_addr(int index)
+{
+#if CONFIG_APOLLOLAKE
+	if (index < DAI_NUM_SSP_BASE)
+		return TIMESTAMP_BASE + TS_I2S_TSCC(index);
+	else
+		return TIMESTAMP_BASE + TS_I2SE_TSCC(index);
+#else
+	return TIMESTAMP_BASE + TS_I2S_TSCC(index)
+#endif
+}
+
+static inline uint64_t io_reg_read_64(uint32_t addr)
+{
+	return (uint64_t)io_reg_read(addr) +
+		(((uint64_t)io_reg_read(addr + 4)) << 32);
+}
+
+static int ssp_ts_config(struct dai *dai, struct sof_timestamp_cfg *cfg)
+{
+	/* Reserved, no settings needed currently */
+	trace_ssp("ssp_ts_config(): SSP=%d, chan=%d",
+		  cfg->index, cfg->dma_chan_index);
+
+	if (cfg->type != SOF_DAI_INTEL_SSP) {
+		trace_ssp_error("ssp_ts_config(): Illegal DAI type");
+		return -EINVAL;
+	}
+
+	if (cfg->index > DAI_NUM_SSP_BASE + DAI_NUM_SSP_EXT - 1) {
+		trace_ssp_error("ssp_ts_config(): ");
+		return -EINVAL;
+	}
+
+	/* TODO, use some existing platform macro */
+	cfg->walclk_rate = 19200000;
```

Comment:
I'd prefer to retrieve this from platform specific macros instead of this magic value but I'm not sure what to use.

Reply:
Is this supposed to be the clock rate for whatever the SSP uses to generate its time stamps?

Reply:
Also, the HW timer retrieve via DAI plus generic sample count can be preserved in ASRC. Or if no desire to wrap a HW timer under DAI there could be other service for ASRC to get the time reference.